src/stream/src/executor/flow_control.rs

```diff
@@ -79,15 +79,23 @@ impl FlowControlExecutor {
                         continue;
                     };
                     if let Some(rate_limiter) = &rate_limiter {
-                        let result = rate_limiter.until_n_ready(n).await;
-                        if let Err(InsufficientCapacity(_max_cells)) = result {
-                            tracing::error!(
-                                "Rate Limit {:?} smaller than chunk cardinality {chunk_cardinality}",
-                                self.rate_limit,
-                            );
+                        let limit = NonZeroU32::new(self.rate_limit.unwrap()).unwrap();
+                        if n <= limit {
+                            // `InsufficientCapacity` should never happen because we have done the check
+                            rate_limiter.until_n_ready(n).await.unwrap();
+                            yield Message::Chunk(chunk);
+                        } else {
+                            // Cut the chunk into smaller chunks
+                            for chunk in chunk.split(limit.get() as usize) {
```

Comment:
I think the logic is not correct when a chunk to be split but the limit rate changes

eg. a chunk of cardinality 10 to be split into 4, 4, 2.
but after yielding the first split chunk of cardinality 4, the limit changes to 3
in current impl, the next chunk be yield is of cardinality 4 rather 3. 

```
- chunk.split(n)
+ chunk.get_first_n_row(n)
```

Reply:
Limit rate changes cannot happen because it must be carried by a barrier, but here it's consuming a data chunk.